api/src/main/java/jakarta/transaction/Transactional.java

```
@@ -158,4 +158,19 @@ public enum TxType {
     @Nonbinding
     public Class[] dontRollbackOn() default {};
 
+    /**
+     * The {@code readOnly} element can be set to indicate that the transaction must be read-only.
+     *  A read-only transaction may fail early if a non-read action is executed
```

I am good with the wording here generally, thank you @beikov but I don't recall the "early" scenario? I think I had the understanding this was all going to get reconciled at commit time? I am not sure I parse the wording of the sentence here though so I might be misunderstanding. After this sentence the wording looks enough for me but I am curious the scenario that is being considered here?

---

>I don't recall the "early" scenario?

I guess I wanted to convey the fact that if a user executes e.g. a SQL `insert` statement within a read-only transaction, then the transaction might fail with a rollback due to that statement execution. It is not guaranteed that the transaction fails though, hence the wording.

---

Oh right, that makes sense to me now in the context of what you are saying now. Maybe "and puts participating resources into read-only mode for this transaction" makes it harder to understand as that clause may seem to further specify the scenario but the scenario is defined well enough (at least in my opinion, given my understanding of your clarification here) by just: "A read-only transaction may fail early if a non-read action is executed"